ShaiRandom/Generators/IEnhancedRandomExtensions.cs

```diff
@@ -250,73 +252,178 @@ public static float MaxFloatOf(this IEnhancedRandom rng, float innerBound, float
+        public static void Shuffle<T>(this IEnhancedRandom rng, IList<T> items, Index startIndex)
+        {
+            int length = items.Count;
+            int start = startIndex.GetOffset(items.Count);
+            length -= start;
+
+            rng.Shuffle(items, start, length);
+        }
+
+        /// <summary>
+        /// Shuffles a section of the given IList in-place pseudo-randomly, using the Fisher-Yates (also called Knuth) shuffle algorithm.
+        /// Only items within the given range will be shuffled.
+        /// </summary>
+        /// <param name="rng" />
+        /// <param name="items">an IList; must be non-null but may contain null items</param>
+        /// <param name="range">Range of items in the list to shuffle</param>
+        /// <typeparam name="T">Type of elements in the list</typeparam>
+        public static void Shuffle<T>(this IEnhancedRandom rng, IList<T> items, Range range)
+        {
+            var (offset, length) = range.GetOffsetAndLength(items.Count);
+            ShuffleUnchecked(rng, items, offset, length);
+        }
+
         /// <summary>
         /// Shuffles a section of the given IList in-place pseudo-randomly, using the Fisher-Yates (also called Knuth) shuffle algorithm.
         /// </summary>
         /// <param name="rng" />
         /// <param name="items">an IList; must be non-null but may contain null items</param>
-        /// <param name="offset">the index of the first element of the IList that can be shuffled</param>
+        /// <param name="start">the index of the first element of the IList that can be shuffled</param>
+        public static void Shuffle<T>(this IEnhancedRandom rng, IList<T> items, int start)
+            => rng.Shuffle(items, start, items.Count - start);
+
+        /// <summary>
+        /// Shuffles a section of the given IList in-place pseudo-randomly, using the Fisher-Yates (also called Knuth) shuffle algorithm.
+        /// </summary>
+        /// <param name="rng" />
+        /// <param name="items">an IList; must be non-null but may contain null items</param>
+        /// <param name="start">the index of the first element of the IList that can be shuffled</param>
         /// <param name="length">the length of the section to shuffle</param>
-        public static void Shuffle<T>(this IEnhancedRandom rng, IList<T> items, int offset, int length)
+        public static void Shuffle<T>(this IEnhancedRandom rng, IList<T> items, int start, int length)
+        {
+            if (start < 0 || start >= items.Count)
+                throw new ArgumentOutOfRangeException(nameof(start), "Index given was out of range");
+
+            if (length < 0 || length > items.Count - start)
+                throw new ArgumentOutOfRangeException(nameof(length), "Index given was out of range");
+
+            ShuffleUnchecked(rng, items, start, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Should we have this kind of inlining marker (annotation?) in more places?

Reply:
TLDR; maybe; but it should probably involve some performance testing.

I have a habit of using it where I explicitly know something is basically designed to only be a method that is forwarded to... But whether it's actually of benefit is probably debatable.  It isn't going to harm anything, but this annotation is really just a "strong hint" to the compiler that you would like the function inlined where possible; if the compiler cannot do so for whatever reason, it won't and everything will still compile.

The C# compiler, in general, appears to usually make fairly good decisions about when to inline things vs when not to; it would have likely inlined calls to this method anyway, although I didn't check the IL.  In many cases for ShaiRandom, the attribute is probably irrelevant anyway; inlining virtual method calls is harder for a lot of practical use cases, so it wouldn't do any good for some of those cases anyway.  I only tend to use it in very obvious cases where I have nothing more than a short math helper function or something, or a case like this where this method is an internal helper only called from other functions that wrap it with tiny additions.